observability/prometheus/prometheus.go

```diff
@@ -31,19 +37,149 @@ func init() {
 }
 
 type observer struct {
+	// URLPrefix is the server's prefix
+	URLPrefix string
+	// observeVars are the vars (:foo) in a url that should be turned into a label
+	// Default values for this via new is []string{":map_name",":layer_name",":z"}
+	observeVars   []string
+	inFlightGauge map[string]prometheus.Gauge
+	counter       map[string]*prometheus.CounterVec
+	// duration is partitioned by the HTTP method and handler. It uses custom
+	// buckets based on the expected request duration.
+	duration map[string]*prometheus.HistogramVec
+	// responseSize has no labels, making it a zero-dimensional
+	// ObserverVec.
+	responseSize map[string]*prometheus.HistogramVec
 }
 
 func New(config dict.Dicter) (observability.Interface, error) {
 	// We don't have anything for now for the config
 	var obs observer
 
+	for _, name := range []string{httpAPI, httpViewer} {
+		if obs.inFlightGauge == nil {
+			obs.inFlightGauge = make(map[string]prometheus.Gauge, 2)
+		}
+		obs.inFlightGauge[name] = prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: name + "_flight_requests",
+			Help: "A gauge of requests currently being served by the wrapped handler.",
+		})
+
+		if obs.counter == nil {
+			obs.counter = make(map[string]*prometheus.CounterVec, 2)
+		}
+		obs.counter[name] = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: name + "_requests_total",
+				Help: "A counter for requests to the wrapped handler.",
+			},
+			[]string{"code", "method"},
+		)
+
+		if obs.duration == nil {
+			obs.duration = make(map[string]*prometheus.HistogramVec, 2)
+		}
+		obs.observeVars, _ = config.StringSlice("variables")
+		if len(obs.observeVars) == 0 {
+			obs.observeVars = []string{":map_name", ":layer_name", ":z"}
+		}
+		durationLabels := []string{"handler", "method"}
+		// duration is partitioned by the HTTP method and handler. It uses custom
+		// buckets based on the expected request duration.
+		obs.duration[name] = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    name + "_duration_seconds",
+				Help:    "A histogram of latencies for requests.",
+				Buckets: []float64{.25, .5, 1, 2.5, 5, 10},
+			},
+			durationLabels,
+		)
+
+		if obs.responseSize == nil {
+			obs.responseSize = make(map[string]*prometheus.HistogramVec, 2)
+		}
+		// responseSize has no labels, making it a zero-dimensional
+		// ObserverVec.
+		obs.responseSize[name] = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name: name + "_response_size_bytes",
+				Help: "A histogram of response sizes for requests.",
+				// TODO(gdey): Do we want to make this configurable? This is the buckets for tiles served up in bytes.
+				Buckets: []float64{512, float64(1 * kb), float64(10 * kb), float64(1 * mb), float64(5 * mb)},
```

Comment:
I don't think we need to make this configurable, but I do think we need a bucket at 500kb rather than 5mb. 500kb is usually a threshold to keep map tiles under. anything over 1mb can really be grouped together.